### DISCUSSION TITLE
Quarto and the European Union General Data Protection Regulation (GDPR)

### DISCUSSION THREAD

**Discussion by bam098** [quarto-dev/quarto-cli#3907] [Q&A] [answered]
I'm new to Quarto and I would like to set up my personal website (incl. blog) using Quarto. However, I still have a question before setting up the website: Does Quarto make any requests to external services? Does Quarto comply to the European Union General Data Protection Regulation (GDPR)?

I just wonder about that, because in Germany there have been cases where several websites used the Google Fonts service and apparently this violated  the European Union General Data Protection Regulation (GDPR)? https://en.wikipedia.org/wiki/Google_Fonts#Privacy_issues

I'm not sure if this is not the right place to ask this kind of question. If it is not, I apologize. I just didn't know where else to ask it.

**Comment by mcanouil** [comment, marked as answer]
It will mostly depends on what you do with Quarto.
Quarto uses seveal open source "tools", listed in https://quarto.org/license.html

It is my belief that Quarto in its default setup does not violate the GDPR, but I might be wrong.


Regarding Google Font, you can download them, and use your local copy.

**Comment by declann** [comment]
Hi,

The Observable runtime is liable to make CDN calls. Some observations I've noted:

- qmd file with only markdown: no ojs cells (nor Julia/other which I amn't using so far): everything in my webpage served locally
- qmd file with a simple {qjs} cell doing `2+2`: everything served locally
- qmd file with `_.range(0,10)`: **results in a call to jsdelivr to load lodash**

**Imports from observablehq *unsurprisingly* lead to calls to api.observablehq.com**. They can be mitigated by hosting locally and loading using FileAttachments (confirm rights with license. I think this approach is OK from OHQ terms but I might be wrong). **A quarto process for this would be smoother than a manual process.**

+It would be better if lodash, and other things that might be fetched from the runtime, can be served locally - for performance and privacy reasons - or at least for this option to be available? Observable runtime **appears** to be quite flexible to make this possible.

(Performance is actually big here despite cache, see my comments in #5411)

Finally, there are other cases where I noticed **calls to Google Fonts**: where at render time I wasn't connected to the internet, so it wasn't possible to download theme fonts locally during the render process, so a link was included. I didn't look for warnings in the CLI output, but my conclusion is *if you want to understand how your Quarto site is behaving to make compliance decisions and to write your privacy policy accurately then I think you need to check how its behaving in practice.*

edit: to not leave this comment on a negative note, we can clearly remote this requirement in relation to the issues noted with development that doesn't seem very major. I think its consistent with how Quarto wants to behave, if we agree on that then I think it becomes a matter of logging some issues and if issues aren't addressed then noting known behavior in the documentation (which is exceptional quality).

edit2: alternatively, only note Quartos known behavior, to facilitate compliance and privacy policy work

Regards,
Declan

**Reply by cscheid** [reply]
> +It would be better if lodash, and other things that might be fetched from the runtime, can be served locally - for performance and privacy reasons - or at least for this option to be available? Observable runtime appears to be quite flexible to make this possible.

I've spent about a dozen hours looking into this and it's my belief that it's "possible", but very hard (anywhere from 100 to 200 hours of work). The relevant repositories are:

- https://github.com/observablehq/runtime/
- https://github.com/observablehq/stdlib/

PRs are always welcome: we host a fork of these repositories in https://github.com/quarto-dev/quarto/tree/main/packages/ojs.

**Reply by declann** [reply]
Thanks for the response. I think I won't be doing this kind of digging up OJS, but I will be interested in some workaround.

Looking further, I can see that calling `d3` results in jsdelivr loads, `Inputs.range` also... There might be workarounds as in overriding _, d3, and Inputs to local copies via FileAttachments, and there might be similar workaround approaches for Quarto more broadly that isn't as messy for the user. I'm sure there are cons to be considered however.

Separate Q on if remote requests should be noted in documentation is still open. If we can't control the calls in particular, then noting to Quarto users (who are devs) that they happen is a good response imo.

**Reply by declann** [reply]
For future reference,

https://observablehq.com/@observablehq/recommended-libraries
This is likely the list that the OHQ standard runtime is liable to download in different notebooks.

Also, the FAQ talks about overriding these libraries in notebooks, so that perhaps doing so is a safe workaround I'll consider.

**Reply by declann** [reply]
> For future reference,
> 
> https://observablehq.com/@observablehq/recommended-libraries This is likely the list that the OHQ standard runtime is liable to download in different notebooks.
> 
> Also, the FAQ talks about overriding these libraries in notebooks, so that perhaps doing so is a safe workaround I'll consider.

So, this was quite messy, but local overrides that are simple copies of the cdn files don't work because there are async require calls inside them.

In a simplified case of inputs.min.js, which requires htl, I got this working by mirroring the structure on jsdelivr and after rendering my project, doing a replacement  of `https://cdn.jsdelivr.net/npm/` for my local reference `http://127.0.0.1:8080/resources/`.

I'm not sure if this will works in itself more widely (maybe scripts are so free to do their own cdn calls, it's probably difficult to make guarantees without reviewing their updates for instance), but it will be something I'll continue to explore.

**Reply by declann** [reply]
+new snag, the [listings](https://quarto.org/docs/websites/website-listings.html) feature adds the following new resources to generated HTML:

```
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
```

Not clear to me why these are needed, but even if I figure that out and/or move to a local approach, then there will still be other features to analyse.

For a specific website/purpose OTOH I think it seems feasible to move away from remote resources - with some work and care. And I still have some temptation to just work with CDNs for their pros and cons likely adding a note to the Privacy Policy (IANAL).

Worth noting that the specific issue above is Quarto and not OJS-related/specific.

**Reply by dragonstyle** [reply]
Those are both supporting MathJax. I expect us to use this approach when rendering a `self-contained` or `embed-resources` HTML document which includes Math. The JS files are quite large, so we link to external sources to reduce the download size of the generated HTML document. You can control this using `self-contained-math: false`.

If you're seeing this in other circumstances, it would be good to know about because I'm not 100% sure why they would be there.

**Reply by declann** [reply]
Thanks for the response, I'll note here if I notice this again in my own use of Quarto.

Do existing options allow me to use a local copy of MathJax? I didn't find something.

Since my website doesn't use MathJax & it isn't included in other pages, right now I'm manually commenting the MathJax/polyfill tags that the listing feature generates, you can see this here:

https://models-on-a-plane.pages.dev/stories/

![image](https://github.com/quarto-dev/quarto-cli/assets/353214/36faaddd-5b30-45df-b44a-dfe6e01b725f)

The website functions incl. listing function works as I intended & without unintentional remote resources & without MathJax - so it's not clear to me why the listing function loads them at all!

Do you know if the polyfill requirement is tied to MathJax, or tied to the listing feature? My guess is its tied to listing, to support some old browsers, and could be slightly more difficult to serve locally.

Overall, unless I'm missing something I think Quarto is missing an option to serve local resources instead of remote ones. This is genuinely a little complicated for OJS (I have a hacky workaround I can note separately in due course), but the problem is wider than that. I think there is a Google Fonts issue also (see https://cbrnr.github.io/blog/removing-google-fonts/ ).

**Reply by dragonstyle** [reply]
The polyfill is also required for MathJax - the listings feature is actually just enabling Math rendering since it doesn't 'know' what the contents of descriptions will be and whether math rendering will be required. I don't see a way to control this explicitly (either as an HTML option or for listings specifically). Both would make nice enhancements.

**Reply by declann** [reply]
The behaviour makes a lot of sense to me now, thanks!

**Reply by declann** [reply]
Just to note a new one, I've started to use Python cells and when I call `ojs_define` require.js and jquery scripts via cloudflare CDN are added to HTML output.

![image](https://github.com/quarto-dev/quarto-cli/assets/353214/67a10bc2-f1e0-4442-8bac-d0c384a997b7)

edit: Also, they don't appear to be used; I can comment out without any impact.

**Comment by declann** [comment]
> Does Quarto make any requests to external services?

In very common cases yes. I'd make the following Quarto suggestions to improve user experience and facilitate flexibility:

- an option for locally-served Google Fonts
- an option for locally-served math resources (alt. or a single option for both of these)
  - incl. the polyfill.io dependency - which is a dynamic response from an external service, but maybe some static local asset will work
- for OJS, control over CDN references (hard to complete a good user experience here)

+There are probably other areas where Quarto makes remote calls I haven't encountered/understood yet.

> Does Quarto comply to the European Union General Data Protection Regulation (GDPR)?

Interpretation and IANAL but this is really each websites responsibility I think?

There isn't something inherently wrong with using Google Fonts, but there are obligations about consent, etc. (but again I am not a lawyer)

However, IMO Quarto should help users understand how generated websites work in terms of resources, and can be better at this.

**Comment by lerchchr** [comment]
Old question, but very similar topic:
[Webkoll](https://webbkoll.5july.net/en) gives for Quarto [this report](https://webbkoll.5july.net/en/results?url=http%3A%2F%2Fquarto.org). Using localStorage is similar to cookies. Is there a way to avoid this with Quarto?

**Reply by cderv** [reply]
Thanks for sharing this report. We'll have a look. Regarding your question about localStorage, it is used in Quarto only to store [grouped tab](https://quarto.org/docs/output-formats/html-basics.html#tabset-groups) so that all tabset on the page are switched together. Code is here:  
https://github.com/quarto-dev/quarto-cli/blob/555a4d1356ef35ecc04144be2cfe184db048d34d/src/resources/formats/html/quarto.js#L803-L894

I don't think this is really a concern regarding GDPR so that you need to avoid this. Though if you don't use group tabset than localStorage won't be used. 

It is not reported by your report but we also use it for storing color scheme value.
https://github.com/quarto-dev/quarto-cli/blob/555a4d1356ef35ecc04144be2cfe184db048d34d/src/resources/formats/html/templates/quarto-html.ejs#L146-L154
Same - I don't think this is a concern regarding GDPR either. 

Thanks a again for the question.

**Reply by lerchchr** [reply]
Thanks for the swift reply. I agree there should be no concern regarding GDPR in terms of the content of localStorage - it could be different form a more legal point of view.
Technically, I stumpled upon this:
If I use grouped tabs, the content of the localStorage "quarto-persistent-tabsets-data" is as expected, e.g. {"language":"Python"} (see screenshot 1)
![localStorage with content](https://github.com/user-attachments/assets/54e82e34-08dc-4e8f-940a-8d051eadb3bd).

But if I don't use (grouped) tabs, "quarto-persistent-tabsets-data" is still stored without any content, ie {} (see screenshot 2)
![localStorage without content](https://github.com/user-attachments/assets/a621276d-44b9-4f47-a244-823bcbb32df7)

Why should it be necessary the store "quarto-persistent-tabsets-data" if there are no (grouped) tabs without any content?
